### PR TITLE
[SPARK-25164][SQL] Avoid rebuilding column and path list for each column in parquet reader

### PR DESCRIPTION
**Pull request by bersprockets**
## What changes were proposed in this pull request?

VectorizedParquetRecordReader::initializeInternal rebuilds the column list and path list once for each column. Therefore, it indirectly iterates 2\*colCount\*colCount times for each parquet file.

This inefficiency impacts jobs that read parquet-backed tables with many columns and many files. Jobs that read tables with few columns or few files are not impacted.

This PR changes initializeInternal so that it builds each list only once.

I ran benchmarks on my laptop with 1 worker thread, running this query:
<pre>
sql("select * from parquet_backed_table where id1 = 1").collect
</pre>
There are roughly one matching row for every 425 rows, and the matching rows are sprinkled pretty evenly throughout the table (that is, every page for column <code>id1</code> has at least one matching row).

6000 columns, 1 million rows, 67 32M files:

master | branch | improvement
-------|---------|-----------
10.87 min | 6.09 min | 44%

6000 columns, 1 million rows, 23 98m files:

master | branch | improvement
-------|---------|-----------
7.39 min | 5.80 min | 21%

600 columns 10 million rows, 67 32M files:

master | branch | improvement
-------|---------|-----------
1.95 min | 1.96 min | -0.5%

60 columns, 100 million rows, 67 32M files:

master | branch | improvement
-------|---------|-----------
0.55 min | 0.55 min | 0%

## How was this patch tested?

- sql unit tests
- pyspark-sql tests

### CODE REVIEW

**Comment by vanzin** [comment]
LGTM. Will leave here for a bit to see if anyone else comments...

**Comment by bersprockets** [comment]
Thanks @vanzin. In my benchmark tests, the tiny degradation (0.5%) in the lower column count cases is pretty consistent, which concerns me a little. I am going to re-run those tests in a different environment and see what happens.

**Comment by bersprockets** [comment]
OK, I reran the tests for the lower column count cases, and the runs with the patch consistently show a tiny (1-3%) improvement compared to the master branch. So even the lower column count cases benefit a little.

**Review by HyukjinKwon — APPROVED** [review]
LGTM

**Review by gatorsmile — COMMENTED** [review]
LGTM

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java [hunk 1]
```diff
@@ -270,21 +270,23 @@ public boolean nextBatch() throws IOException {
   private void initializeInternal() throws IOException, UnsupportedOperationException {
     // Check that the requested schema is supported.
     missingColumns = new boolean[requestedSchema.getFieldCount()];
+    List<ColumnDescriptor> columns = requestedSchema.getColumns();
+    List<String[]> paths = requestedSchema.getPaths();
```
Comment (gatorsmile) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java]: cc @michal-databricks @mswit-databricks

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by bersprockets** [comment]
@cloud-fan @gatorsmile Should we merge this also onto 2.2? It was a clean cherry-pick for me (from master to branch-2.2), and I ran the top and bottom tests (6000 columns, 1 million rows, 67 32M files, and 60 columns, 100 million rows, 67 32M files) from the PR description and got the same results.

**Comment by gatorsmile** [comment]
@bersprockets The risk is pretty small I think. I am fine to backport it to the previous versions. Why 2.2 only?

**Comment by bersprockets** [comment]
@gatorsmile 
>Why 2.2 only?

Only that I forgot that master is already on 2.4. We should do 2.3 as well, but I haven't tested it yet.

Do I need to do anything on my end to get it into 2.2, and once I test, into 2.3?

**Comment by gatorsmile** [comment]
Normally, we do not backport such improvement PRs. However, the risk of this PR is pretty small. I think it is fine. Let me do this.

**Comment by bersprockets** [comment]
@gatorsmile Thanks much!

**Pull request closed without merging**